File: orchestrator/spec.py
import os
import yaml
from typing import Any

Spec = dict[str, Any]


def load(path: str, extra_specs: list[str] | None = None) -> Spec:
    project: Spec | None = None
    files = [path] + (extra_specs or [])
    for f in files:
        if not os.path.exists(f):
            raise FileNotFoundError(f"Spec file not found: {f}")
        with open(f) as fh:
            data: Spec = yaml.safe_load(fh)
        if data is None:
            raise ValueError(f"Empty spec file: {f}")
        if project is None:
            project = data
        else:
            project["milestones"].extend(data.get("milestones", []))
    if project is None:
        raise FileNotFoundError("No spec files provided")
    for key in ("project", "milestones"):
        if key not in project:
            raise ValueError(f"Spec missing required key: {key!r}")
    return project


def get_milestone(project: Spec, name: str) -> Spec | None:
    for ms in project.get("milestones", []):
        if ms.get("name") == name:
            return ms
    return None


def preflight(project: Spec) -> list[str]:
    errors: list[str] = []
    names = set()
    for ms in project.get("milestones", []):
        name = ms.get("name")
        if not name:
            errors.append("Milestone without name")
            continue
        if name in names:
            errors.append(f"Duplicate milestone name: {name}")
        names.add(name)
        tasks = ms.get("tasks", [])
        if not tasks:
            errors.append(f"Milestone {name!r} has no tasks")
    for ms in project.get("milestones", []):
        for dep in ms.get("depends_on", []):
            if dep not in names:
                errors.append(f"Milestone {ms['name']!r} has unknown dependency: {dep!r}")

    return errors

File: orchestrator/internal/spec/spec.go
package spec

import (
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

type Project struct {
	Name        string `yaml:"name"`
	Description string `yaml:"description,omitempty"`
}

type TaskSpec struct {
	Name   string `yaml:"name"`
	Prompt string `yaml:"prompt"`
}

type Milestone struct {
	Name      string     `yaml:"name"`
	Spec      string     `yaml:"spec,omitempty"`
	DependsOn []string   `yaml:"depends_on,omitempty"`
	Tasks     []TaskSpec `yaml:"tasks"`
	Verify    []string   `yaml:"verify,omitempty"`
}

type Spec struct {
	Project    Project     `yaml:"project"`
	Milestones []Milestone `yaml:"milestones"`
}

func Load(path string, extraSpecs ...string) (*Spec, error) {
	var spec Spec
	files := append([]string{path}, extraSpecs...)
	for i, f := range files {
		data, err := os.ReadFile(f)
		if err != nil {
			return nil, err
		}
		var partial Spec
		if err := yaml.Unmarshal(data, &partial); err != nil {
			return nil, fmt.Errorf("%s: %w", f, err)
		}
		if i == 0 {
			spec = partial
		} else {
			spec.Milestones = append(spec.Milestones, partial.Milestones...)
		}
	}
	if spec.Project.Name == "" {
		return nil, fmt.Errorf("spec missing required key: project")
	}
	if len(spec.Milestones) == 0 {
		return nil, fmt.Errorf("spec missing required key: milestones")
	}
	return &spec, nil
}

func GetMilestone(s *Spec, name string) *Milestone {
	for i := range s.Milestones {
		if s.Milestones[i].Name == name {
			return &s.Milestones[i]
		}
	}
	return nil
}

func Preflight(s *Spec) []string {
	var errors []string
	names := make(map[string]bool)
	for _, ms := range s.Milestones {
		if ms.Name == "" {
			errors = append(errors, "milestone without name")
			continue
		}
		if names[ms.Name] {
			errors = append(errors, fmt.Sprintf("duplicate milestone name: %s", ms.Name))
		}
		names[ms.Name] = true
		if len(ms.Tasks) == 0 {
			errors = append(errors, fmt.Sprintf("milestone %q has no tasks", ms.Name))
		}
	}
	for _, ms := range s.Milestones {
		for _, dep := range ms.DependsOn {
			if !names[dep] {
				errors = append(errors, fmt.Sprintf("milestone %q has unknown dependency: %q", ms.Name, dep))
			}
		}
	}
	return errors
}
